Refactor: extract out processing of Telnet DO/DONT/WILL/WONT/SB

SlySven opened the pull request:
This should make the behaviour more consistent across the various options - and importantly IMHO make the STATUS (5) option work better because the previous code was not setting/resetting the elements in
the `(bool)` `myOptionState[256]`, `hisOptionState[256]`, `announcedState[256]`, `heAnnouncedState[256]` and `triedToEnable[256]` arrays for several options that we do process - with the result I suspect that we would report them erronously back to the Game Server should it actually ask what the settings were.  I also suspect that some individual `bool` members of the `Host` or `cTelnet` class that track whether an option is in use or not may be redundant, because we could/should instead consider the value in one of those arrays:
* `Host::mServerMXPenabled` ==> `cTelnet::hisOptionState[OPT_MXP]`
* `cTelnet::mMCCP_version_1` ==> `cTelnet::hisOptionState[OPT_COMPRESS]`
* `cTelnet::mMCCP_version_2` ==> `cTelnet::hisOptionState[OPT_COMPRESS2]`
* `cTelnet::enableATCP` ==> `cTelnet::hisOptionState[OPT_ATCP]`
* `cTelnet::enableGMCP` ==> `cTelnet::hisOptionState[OPT_GMCP]`
* `cTelnet::enableMSSP` ==> `cTelnet::hisOptionState[OPT_MSSP]`
* `cTelnet::enableMSP` ==> `cTelnet::hisOptionState[OPT_MSP]`
* `cTelnet::enableChannel102` ==> `cTelnet::hisOptionState[OPT_102]`

Refactoring this part of the `cTelnet` class has also made me aware that we may not be fully handling the fact that sub-option negotiation is directional - in that there is a separate option for each feature that may be enabled or disabled in each direction independently and separately.  Some options - like TTYPE are only activated at Mudlet (the client's) end and some like MCCP/MCCP2 are only activated at the Game Server's end {though MCCP**3** is Scandum's flawed - IMHO - attempt to do the same sort of compression in the other direction - when really it should just be negotiated by the Server **sending** `IAC WILL MCCP2`and the client responding `IAC DO`/`DONT MCCP2` as appropriate!} Anyhow our current `sysProtocolEnabled` and `sysProtocolDisabled` events do not indicate which direction the protocol has been changed for...

~~The reason for this PR is so that I can clean up this section of code before I look towards trying to implement CHARSET negotiation support.~~ Already done in a basic manner (no negotiation of what Mudlet is set for, instead it adopts what Server suggests

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

I would like to get confirmation that this PR does not make any worse (and maybe improves) the negotiation of the following options - we'll tick 'em off when confirmed that they still work:

- [ ] Echo (1)
- [ ] Status (5)
- [ ] TTYPE (24)
- [ ] EOR (25)
- [ ] NAWS (31)
- [ ] MSDP (69)
- [ ] MSSP (70)
- [ ] MCCP2 (86)
- [ ] MSP (90)
- [ ] MXP (91)
- [ ] Aardwulf (102)
- [ ] ATCP (200)
- [ ] GMCP (201)

SlySven committed:
Refactor: extract out processing of Telnet DO/DONT/WILL/WONT/SB

This should make the behaviour more consistent across the various
options - and importantly IMHO make the STATUS (5) option work better
because the previous code was not setting/resetting the elements in
the `(bool)` `myOptionState[256]`, `hisOptionState[256]`,
`announcedState[256]`, `heAnnouncedState[256]` and `triedToEnable[256]`
arrays for several options that we do process - with the result I
suspect that we would report them erronously back to the Game Server
should it actually ask what the settings were.  I also suspect
that some individual `bool` members of the `Host` or `cTelnet` class
that track whether an option is in use or not may be redundant, because
we could/should instead consider the value in one of those arrays:
* `Host::mServerMXPenabled` ==> `cTelnet::hisOptionState[OPT_MXP]`
* `cTelnet::mMCCP_version_1` ==> `cTelnet::hisOptionState[OPT_COMPRESS]`
* `cTelnet::mMCCP_version_2` ==> `cTelnet::hisOptionState[OPT_COMPRESS2]`
* `cTelnet::enableATCP` ==> `cTelnet::hisOptionState[OPT_ATCP]`
* `cTelnet::enableGMCP` ==> `cTelnet::hisOptionState[OPT_GMCP]`
* `cTelnet::enableMSSP` ==> `cTelnet::hisOptionState[OPT_MSSP]`
* `cTelnet::enableMSP` ==> `cTelnet::hisOptionState[OPT_MSP]`
* `cTelnet::enableChannel102` ==> `cTelnet::hisOptionState[OPT_102]`

Refactoring this part of the `cTelnet` class has also made me aware
that we may not be fully handling the fact that sub-option negotiation
is directional - in that there is a separate option for each feature
that may be enabled or disabled in each direction independently and
separately.  Some options - like TTYPE are only activated at Mudlet
(the client's) end and some like MCCP/MCCP2 are only activated at the
Game Server's end {though MCCP**3** is Scandum's flawed - IMHO -
attempt to do the same sort of compression in the other direction -
when really it should just be negotiated by the Server **sending**
`IAC WILL MCCP2`and the client responding `IAC DO`/`DONT MCCP2` as
appropriate!} Anyhow our current `sysProtocolEnabled` and
`sysProtocolDisabled` events do not indicate which direction the
protocol has been changed for...

The reason for this PR is so that I can clean up this section of code
before I look towards trying to impliment CHARSET negotiation support.

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented on src/ctelnet.cpp:
```diff
@@ -929,17 +929,199 @@ std::pair<QString, int> cTelnet::getConnectionInfo() const
     }
 }
 
+void cTelnet::processTelnetDo(const std::string& command)
```
Unfortunately the argument to this and other *existing* methods in this class (specifically `processCommand(...)`) was given the same name as a member of the class - so I renamed the latter so that it won't be hidden/masked by the use of this name for this argument... :stuck_out_tongue_winking_eye:

vadi2 commented:
Looks like a good change! It doesn't want to compile with `clang` though. Neat how the errors are shown in-line in the `Files changed` view thanks to the [run-cmake](https://github.com/marketplace/actions/run-cmake) action.

SlySven committed:
BugFix: include needed static_cast<unsigned char>(...)s

The value concerned is being used to iterate through an array (of size 256)
so it needed to be a `size_t` type but in the particular uses it needs to
be converted to an `unsigned char` in a position where it is too wide
without the explicit cast...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Refactor_extractOutProcessingOfTelnetDoDontWillWontSb

Conflicts resolved in:
* src/ctelnet.cpp

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
BugFix: use correct cast type this time!

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Revise: remove warnings from combining things in std::string::append(...)

In this PR I was trying to combine as many single std::string::append(...)s
(in the operator+() form) but I pushed it to far in a couple of places...

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven committed:
Merge branch 'development' into Refactor_extractOutProcessingOfTelnetDoDontWillWontSb

Needed to include latest Qt 5.12.8 in some CI builds which were failing
because they were set to 5.12.7 .

Signed-off-by: Stephen Lyons <slysven@virginmedia.com>

SlySven commented:
:facepalm: Accidentally deleted this branch whilst cleaning up a local repo mess - (pushed out a load of branches to GitHub rather than a local machine).

vadi2 commented:
Please re-open when ready, closing for now.
